document/core/valid/conventions.rst

```diff
@@ -24,6 +24,213 @@ That is, they only formulate the constraints, they do not define an algorithm.
+
+It is hence an invariant of the semantics that all :ref:`recursive types <syntax-rectype>` occurring in defined types are :ref:`rolled up <aux-roll-rectype>`.
+
+
+.. index:: ! substitution
+.. _type-subst:
+.. _notation-subst:
+
+Conventions
+...........
+
+* :math:`t[x^\ast \subst \X{dt}^\ast]` denotes the parallel *substitution* of :ref:`type indices <syntax-typeidx>` :math:`x^\ast` with :ref:`defined types <syntax-deftype>` :math:`\X{dt}^\ast` in type :math:`t`, provided :math:`|x^\ast| = |\X{dt}^\ast|`.
+
+* :math:`t[(\REC~i)^\ast \subst \X{dt}^\ast]` denotes the parallel *substitution* of :ref:`recursive type indices <syntax-rectypeidx>` :math:`(\REC~i)^\ast` with :ref:`defined types <syntax-deftype>` :math:`\X{dt}^\ast` in type :math:`t`, provided :math:`|(\REC~i)^\ast| = |\X{dt}^\ast|`.
+
+* :math:`t[\subst \X{dt}^\ast]` is shorthand for the substitution :math:`t[x^\ast \subst \X{dt}^\ast]`, where :math:`x^\ast = 0 \cdots (|\X{dt}^\ast| - 1)`.
+
+
+.. index:: recursive type, defined type, sub type, ! rolling, ! unrolling, ! expansion, type equivalence
+.. _aux-roll-rectype:
+.. _aux-unroll-rectype:
+.. _aux-roll-deftype:
+.. _aux-unroll-deftype:
+.. _aux-expand-deftype:
+
+Rolling and Unrolling
+~~~~~~~~~~~~~~~~~~~~~
+
+In order to allow comparing :ref:`recursive types <syntax-rectype>` for :ref:`equivalence <match-deftype>`, their representation is changed such that all :ref:`type indices <syntax-typeidx>` internal to the same recursive type are replaced by :ref:`recursive type indices <syntax-rectypeidx>`.
+
+.. note::
+   This representation is independent of the type index space,
+   so that it is meaningful across module boundaries.
+   Moreover, this representation ensures that types with equivalent recursive structure are also syntactically equal,
+   hence allowing a simple equality check on (closed) types.
+   It gives rise to an *iso-recursive* interpretation of types.
+
+The representation change is performed by two auxiliary operations on the syntax of :ref:`recursive types <syntax-rectype>`:
+
+* *Rolling up* a recursive type :ref:`substitutes <notation-subst>` its internal :ref:`type indices <syntax-typeidx>` with corresponding :ref:`recursive type indices <syntax-rectypeidx>`.
+
+* *Unrolling* a recursive type :ref:`substitutes <notation-subst>` its :ref:`recursive type indices <syntax-rectypeidx>` with the corresponding :ref:`defined types <syntax-deftype>`.
+
+These operations are extended to :ref:`defined types <syntax-deftype>` and defined as follows:
+
+.. math::
+   \begin{array}{@{}llll@{}}
+   \rollrt_x(\TREC~\subtype^\ast) &=& \TREC~(\subtype[(x + i)^\ast \subst (\REC~i)^\ast])^\ast \\
+   &&& (\iff i^\ast = 0 \cdots (|\subtype^\ast| - 1)) \\
+   \unrollrt(\TREC~\subtype^\ast) &=& \TREC~(\subtype[(\REC~i)^\ast \subst ((\TREC~\subtype^\ast).i)^\ast])^\ast \\[2ex]
+   &&& (\iff i^\ast = 0 \cdots (|\subtype^\ast| - 1)) \\
```

Comment:
indenting here seems messy - it makes sense to put the side-condition on its own line if the previous line is too long to fit it alongside, but the the `&&&` then indents the side-condition past the end of the previous line, creating a scroll bar anyway.